central/group/service/service_impl.go

```diff
@@ -143,28 +146,61 @@ func (s *serviceImpl) DeleteGroup(ctx context.Context, props *storage.GroupPrope
 
 // Helper function that does a diff between two sets of groups and comes up with needed mutations.
 func diffGroups(previous []*storage.Group, required []*storage.Group) (removed []*storage.Group, updated []*storage.Group, added []*storage.Group) {
-	previousByProps := make(map[string]*storage.Group)
+	// This will hold all previous groups mapped by their properties and the rolename. It will later on be used
+	// to determine whether one of the to-be-created groups is required or already covered by one of the existing
+	// groups, iff the properties and role match.
+	groupsByPropsAndRole := make(map[string]struct{})
+
+	previousByID := make(map[string]*storage.Group)
+
 	for _, group := range previous {
-		previousByProps[string(serialize.PropsKey(group.GetProps()))] = group
+		previousByID[group.GetProps().GetId()] = group
+		groupsByPropsAndRole[string(serialize.PropsKey(group.GetProps()))+group.GetRoleName()] = struct{}{}
 	}
-	requiredByProps := make(map[string]*storage.Group)
+
+	requiredByID := make(map[string]*storage.Group)
+	// This will hold all, seemingly to-be-created, groups which do not have an ID set yet.
+	requiredWithoutID := make([]*storage.Group, 0, len(required))
+
 	for _, group := range required {
-		requiredByProps[string(serialize.PropsKey(group.GetProps()))] = group
+		// Need to deal with groups that do not have an ID yet, since they are potentially new groups.
+		if group.GetProps().GetId() == "" {
+			requiredWithoutID = append(requiredWithoutID, group)
+			continue
+		}
+		requiredByID[group.GetProps().GetId()] = group
+		groupsByPropsAndRole[string(serialize.PropsKey(group.GetProps()))+group.GetRoleName()] = struct{}{}
```

Comment:
So, we have three types of groups we have to deal with in `diffGroups`. The `groupsByPropsAndRole` will ensure that we dedupe newly added groups if they have the same properties + rolename as existing ones.
We need to add the required groups which are _not_ new groups to the map, ensuring that we also cover updates of existing groups that may turn out to dedupe newly added ones.

Taking the following example:
```
required groups:
- {"role": a", "props": {"id": "something1", "authProviderID": "something", "key": "email"}}
- {"role": a", "props": {"authProviderID": "something", "key": "email"}}
```
By adding the existing required group to `groupsByPropsAndRole` we ensure we don't create the second group, which would be a duplicate of the first one.

Since we are talking about it, I've added a comment explaining the thoughts behind this + a test that will handle this specifically.